Add support for externally hosted Grafana

jackgill opened the pull request:
Currently Linkerd assumes that the Grafana instance is hosted
on-cluster. Some users would like to use externally hosted Grafana
instances, such as Grafana Cloud or AWS Managed Grafana.

In general users will have multiple Linkerd clusters with dashboards
in the same Grafana workspace, so we need to introduce a prefix for the
Grafana dashboard UID's so that they remain unique. This PR adds
two new viz config values, grafana.uidPrefix and grafana.externalUrl.
When grafana.uidPrefix is set, it will insert the user-supplied prefix
in the URL's for the Grafana dashboards. When grafana.externalUrl is
set, its value will be used for the links to Grafana dashboards instead
of using the grafana reverse proxy.

Fixes #7429

Signed-off-by: Jack Gill <jack.gill@elationhealth.com>

------------------------------------------------------

@alpeb I am submitting this as a draft PR as I suspect it needs some refinement based on your feedback. There are a couple things I'm uncertain about:

1. I was trying to minimize the number of parameters which need to be passed through the code base, so I re-used `grafana.url` instead of introducing `grafana.externalUrl`. That means that the new behavior is controlled by `grafana.uidPrefix`, which might not be obvious based on its name. Do you think it is worth adding another parameter, such as `grafana.externalUrl` or `grafana.isExternal`?
2. I included a Python script to update the dashboard definitions, but I'm not sure if it belongs in this repo or should be re-implemented in Go since Python doesn't seem to be used anywhere else in the project. Maybe it should be in the linkerd-examples repo, or maybe it shouldn't be included at all and we just leave it up to the users to update the dashboard definitions. Let me know what you think.

I also haven't included any documentation updates yet, I thought it would be best to wait to do that until the implementation is finalized.

alpeb reviewed (commented):
This is great, thanks @jackgill !

Good call on the optional prefix :+1: 

Currently we have this already in viz' `values.yaml`, which I think you missed:
```yaml
grafana:
  # -- url of a Grafana instance with reverse proxy configured, used by the
  # Linkerd viz web dashboard to provide direct links to specific Grafana
  # dashboards. See the [Linkerd
  # documentation](https://linkerd.io/2/tasks/grafana) for more information
  url:
```
I think the clearest would be to add an `externalUrl` inside that same section, along with `uidPrefix`. And so the behavior in `GrafanaLink.jsx` will depend on whether `grafana.url` or `grafana.externalUrl` is set (you could add some helm validation in `web.yaml` checking that both values aren't set at the same time). That behavior would then be independent of `grafana.uidPrefix` (which would only be considered if `grafana.externalUrl` is set). This also means only `grafanaPrefix` would have to be passed around, and not `grafana` (if `grafana.externalUrl` is not set, we use the same link as before, with `/grafana/` hard-coded at the beginning).

As for the python script, when importing a dashboard through Grafana Cloud for example, one can change the UID on the import screen. The datasource is not modifiable, but I believe there's a better way to parametrize this. I'll get back to you later on this one (also on how to deal with the script tags).

I hope all this made sense, let me know if you have more questions! :slightly_smiling_face:

jackgill commented:
@alpeb thanks for the feedback! I've made the following changes:

* Removed Python script to generate dashboard definitions
* Removed duplicate definition of `granafa.url` (not sure how I missed that, but I did!)
* Added `grafana.externalUrl` parameter
* Added Helm validation that `grafana.url` and `grafana.externalUrl` are not both set
* Added JS unit test case for `<GrafanaLink>` when used with `grafana.externalUrl`

Let me know if you think any more changes are needed!

alpeb commented on viz/charts/linkerd-viz/values.yaml:
```diff
@@ -54,7 +54,7 @@ enablePSP: false
 # -- url of external prometheus instance
 prometheusUrl: ""
-
+  
```
Looks like this space slipped through

jackgill replied:
Good catch, this has been removed.

alpeb commented on viz/charts/linkerd-viz/values.yaml:
```diff
@@ -349,6 +349,12 @@ grafana:
   # dashboards. See the [Linkerd
   # documentation](https://linkerd.io/2/tasks/grafana) for more information
   url:
+  # -- url of a Grafana instance hosted off-cluster. Can not be set if
+  # grafana.url is set. The reverse proxy will not be used for this URL.
```
Can you complete the comment above `url` saying also that it can't be set if `externalUrl` is also set? And while you're at it, could you also specify in that same coment that the `url` setting is specific for in-cluster Grafana instances?

jackgill replied:
Yep, comment has been updated.

alpeb commented on web/app/js/components/GrafanaLink.jsx:
```diff
@@ -3,30 +3,56 @@ import React from 'react';
+    // grafanaPrefix is used for externally hosted Grafana instances, which do not use the /grafana proxy.
+    // When dashboards for multiple Linkerd instances are deployed to the same Grafana host,
+    // the dashboard UID needs a user-specified prefix to be unique.
+    link = `${grafanaExternalUrl}/d/${grafanaPrefix}-linkerd-${resource}?var-${resource}=${name}`;
```
If `grafanaPrefix` is empty the uid will get prefixed with `-`. I think if people want to separate the prefix with a dash, they should include it in `grafanaPrefix`.
```suggestion
    link = `${grafanaExternalUrl}/d/${grafanaPrefix}linkerd-${resource}?var-${resource}=${name}`;
```

jackgill replied:
Good idea, I have removed the `-`

alpeb reviewed (commented):
Thanks for the quick turnover :+1: :+1: 

I just have some small comments below. One thing I think is missing is not creating the grafana reverse proxy when `externalUrl` is set. That means skipping the `handleAllOperationsForPath()` call in `server.go` and not setting the `grafanaProxy` inside `NewServer()`.

OTOH, I've pushed #7603 which handles the parametrization of the datasource and fixes the script tag problem.

jackgill commented:
@alpeb glad to hear about  #7603, that looks like a great solution!

I've made these changes to address your comments:
* Removed extraneous whitespace after `prometheusUrl`
* Updated comment for `grafana.url` to indicate that it is only for in-cluster Grafana instances, and cannot be set if `grafana.externalUrl` is set
* Removed hyphen after `uidPrefix` in grafana link
* grafana reverse proxy will not be used if `grafana.externalUrl is set`

I also updated the commit message (and PR description) to reflect the changes since the initial version.

Happy to take another pass at this if there's anything else needed!

jackgill commented:
Good call, logic to strip trailing slash, if present, from `grafana.externalUrl` has been added to `GrafanaLink.jsx`.

alpeb reviewed (approved):
Thanks again @jackgill , this is good to go from my perspective :1st_place_medal:

Pothulapati commented on web/app/js/components/GrafanaLink.jsx:
```diff
@@ -3,30 +3,62 @@ import React from 'react';
 
-const GrafanaLink = function({ PrefixedLink, name, namespace, resource }) {
+const GrafanaLink = function({ PrefixedLink, name, namespace, resource, grafanaExternalUrl, grafanaPrefix }) {
   let link = `/grafana/d/linkerd-${resource}?var-${resource}=${name}`;
```
nit: How about we attach `linkerd-${resource}?var-${resource}=${name}` after the if condition below? so as to not repeat this twice i.e inside the if and before.

jackgill replied:
@Pothulapati good idea, I have pushed a change to implement your suggestion.

Pothulapati commented on web/app/js/components/GrafanaLink.test.jsx:
```diff
@@ -47,4 +47,25 @@ describe('GrafanaLink', () => {
     expect(href).toContain(expectedVarNameStr);
   });
+
+  it('Use grafana.externalUrl', () => {
```
:+1:

Pothulapati reviewed (approved):
This is awesome work! Thanks for taking this up!

(Will merge this once the nit is addressed)

Pothulapati commented:
Thanks @jackgill for the awesome work! 👏🏼

jackgill commented:
@alpeb @Pothulapati thanks for your help and advice in getting this done!
